### ISSUE TITLE
Question: Willing to Accept More Curses-Like Features?

### ISSUE THREAD

**Issue by qbradq** [agauniyal/rang#51]
There is a community of people who still develop applications with the ncurses library. Most of these applications do not require the full feature set of ncurses, only require a few basic features. Would you be willing to accept pull requests for the below feature set or do you feel they fall outside the scope of Rang?

```C++
// Hide the cursor
std::cout << rang::control::hideCursor;

// Clear the screen
std::cout << rang::clear;

// Discover terminal dimensions
int w, h;
rang::dimensions(w, h);

// Position cursor
std::cout << rang::pos((w - 20) / 2, 2) << "How many widgets? " << std::endl;

// Get buffered input
string linebuf;
rang::cin >> linebuf;

// Get unbuffered input
std::cout << rang::pos((w - 20) / 2, 4) << "Press any key to exit" << std::endl;
while(true) {
    rang::key key;
    rang::cin >> key;
    if(key == 'q' || key == 'Q') {
        break;
    }
}

// Convenience method for clearing the terminal and resetting all options to normal
std::cout << rang::control::reset;
```

**Comment by agauniyal** [comment]
@qbradq as long as they are in their separate namespace, why not? Those would be great additions to the library and still preserve the minimalism as you say.

My only concern would be their compatibility with sheer number of terminals and of course with windows 10 cmd too. But as long as calling those methods does nothing on non supported terms, we're doing fine. eg calling `rang::style::blink` does nothing on non-supported terms and no garbage is produced either.

Also since I'm not so experienced with using ncurses lib yet, I can make you a contributor of this repo too so you can test out code much faster.

**Comment by qbradq** [comment]
I'll give you a warning up front: I haven't coded in C++ in 20 years and I'm only just getting back into things. Can you help me understand what you mean by a separate namespace? Do you mean a separate wrapping namespace like rang_implementation or a separate client-facing namespace like rang? Also I can test things just fine in my fork. Don't worry about that :)

**Comment by agauniyal** [comment]
@qbradq it's much simpler. `rang_implementation` is for things user don't want to use directly as in with their IDE autocompletion. But other than that, when you'd be introducing a bunch of new methods/objects and if they share some concept, put them in the same namespace eg your example demonstrated - `rang::control::hideCursor;` and `rang::control::reset;` both in same control namespace. but if we put same `hideCursor` inside say `rang::fg`, it will give some other meaning - `rang::fg::hideCursor;`

And please do ask if any support from my side is required, I'll be happy to provide it 😄 .

**Comment by lefticus** [comment]
I would be interested in helping with this also. I was looking initially at starting with the `pos()` code https://msdn.microsoft.com/en-us/library/system.console.setcursorposition(v=vs.110).aspx?f=255&MSPPError=-2147217396&cs-save-lang=1&cs-lang=cpp#code-snippet-1 

@qbradq have you started working on this at all?

**Comment by AxelStrem** [comment]
I would love these features too - ready to help if there's any need

**Comment by qbradq** [comment]
@lefticus I am working on it now. I've been buried at work for the past several days but I finally got that issue resolved. Plus I've got some time off coming :)

Thank you both for offering to help. Feel free to work on what you like. I am starting with positional output. Perhaps one of you could start on the input portion? I was thinking we'd probably want to separate the input handling into a separate header file to keep the compile-time overhead down for those that don't need or want input handling. Thoughts @agauniyal ?

**Comment by agauniyal** [comment]
@qbradq is compile time overhead large enough for this change? As far as I know most people use build systems(or are moving to use one) and most of the project isn't recompiled each time a new change is introduced. Another point is the ease of use this lib provides, just drop the header file and use it. Plus the work this lib does wouldn't be spread across 100s of files that's for sure.

So I would say let's keep it in a single file for some time and test this out 😄 . I'm still open to further feedback though.

**Comment by agauniyal** [comment]
Features - 

 - Hide cursor
    [ ] vt100
    [ ] windows < 10

 - Clear the screen
    [ ] vt100
    [ ] windows < 10

 - Discover Terminal Dimensions
    [ ] vt100
    [ ] windows < 10

 - Position Cursor
    [ ] vt100
    [ ] windows < 10

 - Get buffered input
    [ ] vt100
    [ ] windows < 10

 - Get unbuffered input
    [ ] vt100
    [ ] windows < 10

 - Reset Terminal
    [ ] vt100
    [ ] windows < 10

**Comment by qbradq** [comment]
@agauniyal Input handling could get pretty large. There's a lot of variance between terminals that we could run into. And for some use cases like a terminal-only game terminal output can be spread out all over the place. However I do agree that it's better to wait until a problem presents itself before trying to solve it.

**Comment by agauniyal** [comment]
@qbradq if the situation demands, I won't disagree to splitting up input methods into `rang-input.hpp`.

**Comment by qbradq** [comment]
I realize now that the test program for some of the conditional output will require user interactivity. So I'm going to work on the input side of things first.

**Comment by agauniyal** [comment]
@AxelStrem maybe you can help with 'hide cursor' and 'clear screen' issue since they seem to be well supported in vt100(linux and win >=10) machines but I've no idea with win < 10 versions of cmd. I've linked info in this comment - https://github.com/agauniyal/rang/issues/51#issue-196221739

**Comment by qbradq** [comment]
@agauniyal @AxelStrem For windows < 10 see [Windows Console Functions](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682073(v=vs.85).aspx). Use FillConsoleOutputAttribute() and FillConsoleOutputCharacter() to clear the screen.

**Comment by qbradq** [comment]
@agauniyal What do you think of this interface?
```
rang::Key key;
std::cin
    >> rang::control::blocking
    >> key;
if(key == 'q) { /* Do Stuff */ }
else if (key.isSpecial()) {
    switch(key.which) {
        case rang::Key.F1: help(); break;
    }
}

std::cin
    >> rang::control::nonBlocking
    >> key;
while(key.isEmpty()) { std::cin >> key; }

/* Same processing as before. */

std::string name;
std::cout >> "What is your name? ";
std::cin << name; // Always blocking
std::cout >> "Hello, " >> name >> ", how are you today?" >> std::endl;
```

**Comment by agauniyal** [comment]
@qbradq this looks good to me but again I'm not so experienced with such code. I do have a few ques -

- In `rang::Key::F1`, about these special keys, how many of them would be supported? F1-12? or something else too. And how are these stored and recognized? probably in an enum since switch cases deal with integer values, right?

- what is `rang::Key` basic type here? Or should I frame the ques as, what is it analogous to - a character or a string or something else? I ask this because there are multiple usages of `key` such as `key == 'q'` , `key.isSpecial()`, `switch(key)`, `key.isEmpty()` and with `std::cin` too.

- what does `rang::control::nonBlocking`/ `rang::control::blocking` do behind the scenes?

- How is `rang::control::flush` different from `std::cin`'s method - `clear()` & `ignore()`.

- When should we use unbuffered input vs buffered input? I'm familiar with unbuffered vs buffered output as used with logging/error messages but not so with input.

- would all functionality mentioned, supported with windows and linux both?

Sorry to bug you with lot of ques, I'll need to update the wiki as well so I'm trying to get a prespective about the additions 😄 .

**Comment by agauniyal** [comment]
`ReadConsoleInput` should work on all version of windows, shouldn't it? We just made a distinction between windows >=10 and windows < 10 because windows 10 finally supported ansi escape sequences in cmd, while earlier versions weren't compatible so we had to use native windows methods. So if an existing function which is working on windows 7, it should work in newer versions too unless specified so.

Otherwise everything looks good and I also saw some of the work you did in your branch which looks promising. I would like to see a PR in progress so that we could transfer some of the discussion about input features there and keep this thread for overall discussion about all the features.

**Comment by agauniyal** [comment]
@qbradq you'll be glad to know that terminfo parser has finally landed in `develop` branch. So stuff like hiding cursor, clearing screen are a single line addition to library now 😄 . Though we still need a terminfo parameterized string expansion method before going further.

**Comment by chuyangliu** [comment]
😃 I am glad that I found this library and this issue because I am currently developing games on terminals for research. My games need features below and I have written some simple functions to fulfill it:

- [setCursor()](https://github.com/chuyangliu/Snake/blob/master/src/util/console.cpp#L20)
- [clear()](https://github.com/chuyangliu/Snake/blob/master/src/util/console.cpp#L34)
- [getch()](https://github.com/chuyangliu/Snake/blob/master/src/util/console.cpp#L103) [[Reference]](http://stackoverflow.com/questions/3276546/how-to-implement-getch-function-of-c-in-linux)
- [kbhit()](https://github.com/chuyangliu/Snake/blob/master/src/util/console.cpp#L121) [[Reference]](http://cboard.cprogramming.com/c-programming/63166-kbhit-linux.html)

I am wondering if the last two methods can be added to this library since it seems to fall outside of RANG's scope. Thanks in advance!

**Comment by agauniyal** [comment]
@chuyangliu yes they can be added to the library! There is an undergoing work with terminfo parser which will bring methods like `setCursor()` and `clear()` very soon to all linux terminals and windows already has those functions you've used. The last two functions can be put inside `rang::key` namespace. There is a new release `2.5` coming in 2-3 days which adds support for windows mingw/cygwin/bash and all sorts of terminals, then there will be another release `3.0` with terminfo support maybe in late November or early December where I'll add all these features. You can request more if you need for help with your research and I'll do my best to add them whenever I get time 😄  .

**Comment by chuyangliu** [comment]
@agauniyal Thanks a lot. I am looking forward to it!
